ansible_base/lib/routers/association_resource_router.py

```diff
@@ -0,0 +1,167 @@
+                {
+                    'instances': _('Cannot disassociate these objects because they are not all related to this object: %(non_related_instances)s')
+                    % {'non_related_instances': ', '.join(str(i.pk) for i in non_related_instances)},
+                }
+            )
+
+        try:
+            manager.remove(*related_instances)
+        except AttributeError:
+            # If the field is FK and non-nullable, .remove() won't exist.
+            # https://docs.djangoproject.com/en/dev/ref/models/relations/#django.db.models.fields.related.RelatedManager.remove
+            # "For ForeignKey objects, this method only exists if null=True."
+            # Return something stating that the operation is not supported.
+            raise serializers.ValidationError(
+                {
+                    'instances': _('Cannot disassociate these objects because there must be a related object'),
+                }
+            )
+
+        return Response(status=status.HTTP_204_NO_CONTENT)
+
+    def get_serializer_class(self):
+        association_class = getattr(self, 'association_serializer', None)
+        if self.action in ('associate', 'disassociate'):
+            if association_class is None:
+                raise RuntimeError('You must set association_serializer on the viewset')
+            return association_class
+        return super().get_serializer_class()
+
+
+class ReverseViewMixin(QuerySetMixinBase):
+    child_queryset_attr = 'related_fk'
+
+
+class AssociationResourceRouter(routers.SimpleRouter):
+    def get_method_map(self, viewset, method_map):
+        is_associate_viewset = issubclass(viewset, AssociateMixin)
+        associate_actions = ['associate', 'disassociate', 'list']
+        bound_methods = {}
+        for method, action_str in method_map.items():
+            if hasattr(viewset, action_str):
+                if is_associate_viewset and action_str not in associate_actions:
+                    continue
+                bound_methods[method] = action_str
+        return bound_methods
+
+    def association_serializer_factory(self, related_view):
+        class AssociationSerializer(serializers.Serializer):
+            instances = serializers.PrimaryKeyRelatedField(
+                queryset=related_view.queryset,
+                many=True,
+            )
+
+        return AssociationSerializer
+
+    def register(self, prefix, viewset, related_views={}, basename=None):
+        if basename is None:
+            basename = self.get_default_basename(viewset)
+
+        for related_name, (related_view, fk) in related_views.items():
+            try:
+                for view in [viewset, related_view]:
+                    if getattr(view, 'queryset', None) is None:
+                        logger.error(f"ViewSet {view.__class__} does not have an associated queryset. Unable to add related view {related_name}")
+                        raise MissingQuerySetException()
```

Comment:
https://github.com/ansible/django-ansible-base/blob/devel/test_app/views.py#L12

If it has a queryset, then sure, you can use that. If not, can you use this method to attempt `view.serializer_class.Meta.model`? Do you need anything other than the model?

Reply:
That might be a way to get around some things. We do use the model from the query set. But the associate/disassociate do:
```
instance = self.backward_related_qs.get(pk=kwargs['pk'])
```

But maybe instead of that we could just do a `view.serializer_class.Meta.model.objects.get` instead?

👀

Reply:
That totally worked and gets us around the queryset issue.